File: testconnect.py
import requests
from requests_aws4auth import AWS4Auth
import boto3

# Your OpenSearch domain endpoint
host = 'http://my-local-domain.us-east-1.opensearch.localhost.localstack.cloud:4566'  # For example: 'https://search-mydomain.us-west-1.es.amazonaws.com'

# AWS Region
region = 'us-east-1'  # For example: 'us-west-1'

# AWS Credentials
# AWS Credentials
service = 'es'
session = boto3.Session()
credentials = session.get_credentials()
awsauth = AWS4Auth(credentials.access_key, credentials.secret_key, region, service, session_token=credentials.token)

# Function to create an index
def create_index(index_name):
    url = host + '/' + index_name
    response = requests.put(url, auth=awsauth)

    if response.status_code == 200 or response.status_code == 201:
        print(f"Index '{index_name}' created successfully.")
    else:
        print(f"Error creating index: {response.text}")

# Function to add a document to an index
def add_document(index_name, document, doc_id):
    url = f"{host}/{index_name}/_doc/{doc_id}"
    headers = { "Content-Type": "application/json" }
    response = requests.post(url, auth=awsauth, json=document, headers=headers)

    if response.status_code == 200 or response.status_code == 201:
        print(f"Document added to '{index_name}' successfully.")
    else:
        print(f"Error adding document: {response.text}")

# Example Usage
index_name = "sample-index"
sample_document = {
    "title": "Hello OpenSearch",
    "content": "This is a test document."
}

create_index(index_name)
add_document(index_name, sample_document, doc_id=1)

File: go-opensearch/testConnect.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"io/ioutil"
	"net/http"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	v4 "github.com/aws/aws-sdk-go-v2/aws/signer/v4"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials/stscreds"
	"github.com/aws/aws-sdk-go-v2/service/sts"
)

const (
	host    = "http://my-local-domain.us-east-1.opensearch.localhost.localstack.cloud:4566"
	region  = "us-east-1"
	service = "es"
)

func main() {
	// Load AWS configuration
	cfg, err := config.LoadDefaultConfig(context.TODO(),
		config.WithRegion(region),
	)
	if err != nil {
		panic("configuration error, " + err.Error())
	}

	// Create an STS client
	stsSvc := sts.NewFromConfig(cfg)

	// Retrieve credentials
	// creds := stscreds.NewAssumeRoleProvider(stsSvc, "arn:aws:iam::123456789012:role/my-role")
	//

	credentials := stscreds.NewAssumeRoleProvider(stsSvc, "arn:aws:iam::123456789012:role/my-role")
	cfg.Credentials = credentials

	// Create a static credentials provider
	fmt.Println(credentials.Retrieve(context.Background()))

	// Create a v4 signer

	v4Signer := v4.NewSigner(func(signer *v4.SignerOptions) {})

	// Create a v4 signer

	// Example Usage
	indexName := "sample-index"
	sampleDocument := `{
	        "title": "Hello OpenSearch",
	        "content": "This is a test document."
	    }`

	createIndex(indexName, v4Signer, &cfg)
	addDocument(indexName, sampleDocument, "1", v4Signer, &cfg)
}

func createIndex(indexName string, signer v4.HTTPSigner, cfg *aws.Config) {
	url := host + "/" + indexName
	req, err := http.NewRequest("PUT", url, nil)
	if err != nil {
		panic("request creation failed, " + err.Error())
	}

	signHTTPRequest(req, signer, cfg)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		panic("request failed, " + err.Error())
	}
	defer resp.Body.Close()

	if resp.StatusCode == 200 || resp.StatusCode == 201 {
		fmt.Printf("Index '%s' created successfully.\n", indexName)
	} else {
		body, _ := ioutil.ReadAll(resp.Body)
		fmt.Printf("Error creating index: %s\n", string(body))
	}
}

func addDocument(indexName, document, docID string, signer v4.HTTPSigner, cfg *aws.Config) {
	url := fmt.Sprintf("%s/%s/_doc/%s", host, indexName, docID)
	req, err := http.NewRequest("POST", url, bytes.NewBufferString(document))
	if err != nil {
		panic("request creation failed, " + err.Error())
	}

	req.Header.Set("Content-Type", "application/json")

	signHTTPRequest(req, signer, cfg)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		panic("request failed, " + err.Error())
	}
	defer resp.Body.Close()

	if resp.StatusCode == 200 || resp.StatusCode == 201 {
		fmt.Printf("Document added to '%s' successfully.\n", indexName)
	} else {
		body, _ := ioutil.ReadAll(resp.Body)
		fmt.Printf("Error adding document: %s\n", string(body))
	}
}

func signHTTPRequest(req *http.Request, signer v4.HTTPSigner, cfg *aws.Config) {
	// Sign the request
	creds, err := cfg.Credentials.Retrieve(context.TODO())
	if err != nil {
		panic("request signing failed, " + err.Error())
	}
	err = signer.SignHTTP(context.TODO(), creds, req, "", service, region, time.Now())
	if err != nil {
		panic("request signing failed, " + err.Error())
	}
}
